Fixed validation of the registration process

Hrushon opened the pull request:
Has been changed `registration.html` in regard validation of the fields.
Changed the functions of registration and updation user data in `handlers.py` in terms of providing an opportunity to change invalid data.
Little changed error messages in model validators.

https://www.notion.so/Fix-aebdd0697783441f9fec0531a37c3ae4

Hrushon committed:
fix(bot)!: edited registration form

Has been changed `registration.html` in regard validation of the fields.
Changed the functions of registration and updation user data in `handlers.py` in terms of providing an opportunity to change invalid data.
Little changed error messages in model validators.

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into fix/validation_registration

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -69,12 +69,17 @@ async def register_user(
     """Инициализация формы регистрации пользователя."""
+    query = None
+    text = "Зарегистрироваться в проекте"
+    if update.effective_message.web_app_data:
```
Попробовал запустить у себя локально. Если мы указываем неверную дату рождения (т.е. если ребенок моложе 3-х лет), это условие никогда не выполняется.
Проблема в том, что в методе `web_app_data` мы не попадаем в блок `except`.
На первый взгляд потому, что `UserCreateRequest` валидирует только формат даты, а возраст. Возраст валидируется уже только в `user_service` (который вызывается в `registration_service`).

Т.е., если резюмировать, получается так:
1. Юзер нажимает старт 
2. Открывает форму, вводит неверный возраст
3. Мы вызываем `web_app_data`:
  3.1. получаем `user_data`
  3.2. получаем `user_scheme ` (тут все ок, т.к. валидируется только формат даты)
  3.3. вызываем `registration_service.register_user(user_scheme)`
    3.3.1. вызываем `user_service.__update_or_create_user`
    3.3.2. вызываем `user_service.validate_user_create`
      3.3.2.1. вызываем `user_service.validate_date_of_birth` ловим эксепшен `NotValidValueError`
  3.4. срабатывает блок except, в text записываем `Возраст не может быть менее {settings.MIN_AGE} лет.`
  3.5. отправляем юзеру ответ "Возраст не может быть менее...", удаляем кнопку: `ReplyKeyboardRemove()`
4. Пользователь видит ответ "Возраст не может быть менее...", кнопки нет, что делать дальше - непонятно) Если набрать `/start` - все повторится, форма будет чистая (непредзаполненная)


Вроде бы ничего не упустил)

Нужно это исправить.
Плюс проверь, пожалуйста, другие возможные кейсы у себя)

Hrushon replied:
Не обратил внимание, что валидация даты рождения дополнительно происходит за рамками создания модели запроса на создание пользователя :confused: 
По итогу я правильно понял, что необходимо реализовать перехват исключения `NotValidValueError` при `registration_service.register_user(user_scheme)` и снова повторять эту портянку с вызовом функций обновления или регистрации пользователя?

Hrushon replied:
в общем что-то сделал
не совсем понимаю, что должен видеть пользователь когда приходит сообщение, что смена уже начата и мест нет - должна ли у него маячить кнопка `Пройти регистрацию/Подать заявку на смену` или отдавать кнопку `/start`...

Hrushon committed:
fix(bot)!: edited registration form

Has been changed `registration.html` in regard validation of the fields.
Changed the functions of registration and updation user data in
`handlers.py` in terms of providing an opportunity to change invalid
data.
Little changed error messages in model validators.

Hrushon committed:
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into fix/validation_registration

Hrushon commented:
Может сделать `resize_keyboard=True` в кнопке регистрации? 

Так поаккуратнее вроде смотрится

[<img align="left" alt="Django" width="400px" src="https://user-images.githubusercontent.com/102170248/224771938-1bc50283-1354-4134-a0aa-f7327ae31b85.jpg" />]

Ivan-Skvortsov commented on src/bot/handlers.py:
```diff
@@ -107,23 +118,39 @@ async def web_app_data(update: Update, context: CallbackContext) -> None:
-        text = (
-            " Обновленные данные приняты!\nПроцесс обработки заявок занимает некоторое время - вам придет уведомление."
-        )
+    reply_markup, validation_error = None, False
```
На мой взгляд получилось слишком сложно и запутанно. Предлагаю просто перенести валидацию возраста из метода registrer_user в UserCreateRequest.

Hrushon replied:
@Ivan-Skvortsov 
А что возвращать пользователю в телеграме при `NotValidValueError` в ходе регистрации, когда оказывается, что номер пользователя уже используется с другим телеграм id? Или оставить как есть сейчас?

ADesBiysk replied:
Нужно устранить конфликт, ещё раз проверить на работоспособность код и смержить сегодня или завтра утром. Вечером демонстрация заказчику. 
Далее, в этой ветке можно дальше улучшать код.
С кнопками идея хорошая, но их пока не трогаем. Для них будут отдельное изменение.
Илья, пожалуйста, напиши мне как будет всё проверено и готово.

Hrushon committed:
fix(bot)!: edited registration form

Has been changed `registration.html` in regard validation of the fields.
Changed the functions of registration and updation user data in
`handlers.py` in terms of providing an opportunity to change invalid
data.
Little changed error messages in model validators.
